### PR TITLE
bpo-41972: Use the "Two-Way" algorithm when searching for long substrings

### PR DESCRIPTION
**Pull request by sweeneyde**
https://bugs.python.org/issue41972

### CODE REVIEW

**Review by eamanu — CHANGES_REQUESTED** [review]
IMO you should send the PR when it is ready, to avoid maintain WIP PR for a long  time

**Comment by tim-one** [comment]
> IMO you should send the PR when it is ready,, to avoid maintain WIP PR for a long time

Disagree in this case: this is a change to extremely important core functionality, so needs to be made as easy as possible for others to try. The changes are all in one file that's typically altered less than once per year (for example, the most recent change was an edit to the comments over a year ago - the most recent non-trivial change was in March of 2017).

File: Objects/stringlib/fastsearch.h [hunk 1]
```diff
@@ -160,88 +170,352 @@ STRINGLIB(rfind_char)(const STRINGLIB_CHAR* s, Py_ssize_t n, STRINGLIB_CHAR ch)
...
+        STRINGLIB_CHAR a = needle[suffix + k];
+        STRINGLIB_CHAR b = needle[max_suffix + k];
+        if (inverted ? (a < b) : (b < a)) {
+            // Suffix is smaller, period is entire prefix so far.
+            suffix += k;
+            k = 1;
+            period = suffix - max_suffix;
+        }
+        else if (a == b) {
+            // Advance through the repitition of the current period.
```
Comment (DavidMertz) [inline Objects/stringlib/fastsearch.h]: s/repitition/repetition/

**Comment by sweeneyde** [comment]
The most recent batch of commits added a jump table. 
Between master and this PR now, there are 151 cases slower than master and 463 that faster than master.
The slower cases are at most twice as slow, but the faster cases are often 10-20x faster.
I could add a cutoff to use a simpler algorithm instead, for length 2-3 strings,
but I wanted to get the "purer" data out before making that change.

The benchmark data is here: https://pastebin.com/raw/bzQ4xQgM

**Comment by taleinat** [comment]
@sweeneyde, could you share how exactly you're running the benchmarks? I'd like to check this out myself :)

**Comment by sweeneyde** [comment]
I used this static file containing a bunch of randomly generated strings:
https://gist.github.com/sweeneyde/f77ccf0d25f9c41d41163b11fe64e9b4
This file has:
* One very long haystack
* Needles with lengths ranging from length 1 to 10, then increasing by 50% + 1 up until 100_000
* ~20 strings of each length


Then I ran the benchmarks with this script (`string_benchmarks.py`):
```python
from lots_of_benches import needles, haystack
needles: list[str]
haystack: str

from pyperf import Runner
runner = Runner()

for needle in needles:
    n = len(needle)
    abbrev = needle if n <= 10 else f"{needle[:10]}..."
    runner.timeit(
        name=f"length={n}, value={abbrev}",
        stmt=f"needle in haystack",
        globals=globals(),
    )
```

Run as `python string_benchmarks.py -o whatever.json`

**Comment by pitrou** [comment]
@maartenbreddels if you are interested in string algorithms, this may tick your curiosity.

**Comment by taleinat** [comment]
@sweeneyde, why did you close this?

**Comment by sweeneyde** [comment]
The source code was too heavily based on glibc's, which has a more restrictive license that forbids commercial use.

I opened this as a "clean-room" implementation instead:
https://github.com/python/cpython/pull/22904

**Pull request closed without merging**